README.md

```diff
@@ -1 +1,882 @@
+
+This particular specification, **did:cosmos** applies the IID approach to assets on blockchains made with the Cosmos SDK.
+
+IIDs also introduce a few new features—conformant extensions to the DID Core specification—that provide for privacy-respecting options for the full range of expected token functionality, including Linked Resources, On-chain Service Endpoints, and Accorded Rights. We describe those here, capturing their definition at the time this specification was developed. See the IID specification for current normative definitions.
+
+DID Methods which conform to the IID specification resolve to a DID document representing how to securely interact with a uniquely identified digital asset, within a unique blockchain namespace. Because IIDs are DIDs, software applications that are conformant with the W3C specification will be able to inter-operate with IIDs and IID documents, although some IID-specific features may require additional tooling.
+
+**did:cosmos** DIDs are IIDs intended identify assets on Cosmos application chains.
+
+## Architecture Overview
+
+In the Cosmos ecosystem, **application chains** are sovereign blockchains built using the Cosmos SDK. These application chains are comprised of various modules, which perform different functions. We refer to modules that manage the state of on-chain assets as **asset modules**.
+
+To support **did:cosmos**, three modules must be operational: a **chain registry**, a **namespace registry**, and an **asset module**.
+
+```mermaid
+classDiagram
+    class ChainRegistry
+    ChainRegistry : -chainDescriptors
+    ChainRegistry : +getChainDescriptor(chainspace)
+    ChainRegistry : +setChainDescriptor(chainDescriptor)
+
+    class NamespaceRegistry
+    NamespaceRegistry : -namespaceDescriptors
+    NamespaceRegistry : +getNamespaceDescriptor(namespace)
+    NamespaceRegistry :+setNamespaceDescriptor(namespaceDescriptor)
+
+    class AssetModule
+    AssetModule : -assets
+    AssetModule : +getDidDocument(DID)
+    AssetModule : +createAsset(assetDescriptor)
+    AssetModule : updateAsset(assetDescriptor)
+    AssetModule : deactiveAsset(DID)
+```
+
+**Chain Registry** Maintains a set of chain descriptors which describe how to access particular application chains, indexed by chainspace.
+
+**Namespace Registry** Maintains a set of namespace descriptors which describe how to access particular asset modules, indexed by namespace.
+
+**Asset Module** Maintains on-chain assets, with its own module-specific functions for creating, updating, and deactivating assets. Each Asset Module MUST also support a method for retrieving the DID Document for a given asset, given a DID managed by that module.
+
+The role of each of these modules is further described below.
+
+## did:cosmos Syntax
+
+An example `did:cosmos` method identifier is `did:cosmos:version:chainspace:namespace:unique-id`.
+
+There are four forms of did:cosmos DIDs:
+
+```abnf
+did:cosmos:version:chainspace:namespace:unique-id
```

Comment:
If the version change is a breaking change to how your resolve the DID itself, then you don't have a way to get that DID metadata.

My initial preference was to support a version 0 that explicitly says that resolution starts by looking up the entry in the chain registry github repo, and a later version 1 that would start by querying the on-chain chain registry. However, it is inevitable that the repo and the on-chain registry would, eventually, get out of sync. In that case, which is authoritative?

And more importantly, if old software was built for version 0, it would have no way to know that its algorithm is out of date. It would have no way to know that for this new DID, you should FIRST check the on-chain registry and should not use the github repo. However, with a version in the DID itself, the software can simply verify that the DID it is processing uses a version it understands. Of course, you can't magically get past modifying code to support the new feature, but you can avoid attempting a knowably out-of-date resolution process.

It's these sorts of breaking changes--when extant DIDs in the wild are no longer compatible with the current version--that are addressed by explicitly naming the version in the DID itself.

Reply:
hmmm, I understand the reasoning here but I think in the wild there would be ways to mitigate this such as data migration code and correct proto versioning, but defo best to keep it until we can fully understand the technical implementation details.